File: go-sdk/scratch/let_builder.go
package scratch

import (
	"github.com/gagliardetto/solana-go"
	"github.com/ifx-run/ifx/go-sdk/binding"
	"github.com/ifx-run/ifx/go-sdk/codec"
	"github.com/ifx-run/ifx/go-sdk/expr"
	"github.com/ifx-run/ifx/go-sdk/typed"
)

// LetBuilder plans multi-binding ifx_let with deduped remaining accounts.
type LetBuilder struct {
	scratch      *FrameScratch
	accounts     []typed.AccountMeta
	indexByKey   map[string]int
	bindings     []typed.ScratchValue
}

// FinishResult holds let ix components.
type FinishResult struct {
	Args      codec.LetArgs
	Bindings  []typed.ScratchValue
	Remaining []typed.AccountMeta
	Scratch   *FrameScratch
}

func (b *LetBuilder) AccountIndex(account interface{}) int {
	meta := toLetMeta(account)
	if idx, ok := b.indexByKey[meta.Pubkey]; ok {
		b.accounts[idx] = mergeMeta(b.accounts[idx], meta)
		return idx
	}
	idx := len(b.accounts)
	b.indexByKey[meta.Pubkey] = idx
	b.accounts = append(b.accounts, meta)
	return idx
}

func (b *LetBuilder) push(sv typed.ScratchValue) typed.ScratchValue {
	b.bindings = append(b.bindings, sv)
	return sv
}

func (b *LetBuilder) LetEval(e expr.Node) (typed.ScratchValue, error) {
	sv, err := b.scratch.LetEval(e)
	if err != nil {
		return typed.ScratchValue{}, err
	}
	return b.push(sv), nil
}

func (b *LetBuilder) LetConstU64(n uint64) (typed.ScratchValue, error) {
	sv, err := b.scratch.LetConstU64(n)
	if err != nil {
		return typed.ScratchValue{}, err
	}
	return b.push(sv), nil
}

func (b *LetBuilder) Lamports(account interface{}) (typed.ScratchValue, error) {
	i := b.AccountIndex(account)
	sv, err := b.scratch.PlanAtRemainingIndex(binding.AccountLamports(0), uint8(i))
	if err != nil {
		return typed.ScratchValue{}, err
	}
	return b.push(sv), nil
}

func (b *LetBuilder) DataLen(account interface{}) (typed.ScratchValue, error) {
	i := b.AccountIndex(account)
	sv, err := b.scratch.PlanAtRemainingIndex(binding.AccountDataLen(0), uint8(i))
	if err != nil {
		return typed.ScratchValue{}, err
	}
	return b.push(sv), nil
}

func (b *LetBuilder) LetAccountKey(account interface{}) (typed.ScratchValue, error) {
	i := b.AccountIndex(account)
	sv, err := b.scratch.PlanAtRemainingIndex(binding.AccountKey(0), uint8(i))
	if err != nil {
		return typed.ScratchValue{}, err
	}
	return b.push(sv), nil
}

func (b *LetBuilder) LetConstPubkey(bytes [32]byte) (typed.ScratchValue, error) {
	sv, err := b.scratch.LetConstPubkey(bytes)
	if err != nil {
		return typed.ScratchValue{}, err
	}
	return b.push(sv), nil
}

func (b *LetBuilder) FrameGeneration() (typed.ScratchValue, error) {
	sv, err := b.scratch.LetFrameGeneration()
	if err != nil {
		return typed.ScratchValue{}, err
	}
	return b.push(sv), nil
}

func (b *LetBuilder) FrameIndexCount() (typed.ScratchValue, error) {
	sv, err := b.scratch.LetFrameIndexCount()
	if err != nil {
		return typed.ScratchValue{}, err
	}
	return b.push(sv), nil
}

func (b *LetBuilder) AccountDataSlice(account, expectedOwner interface{}, ty typed.IfxTy, offset uint32) (typed.ScratchValue, error) {
	dataIdx := b.AccountIndex(account)
	ownerIdx := b.AccountIndex(expectedOwner)
	tag, err := typed.ValueTypeTag(ty)
	if err != nil {
		return typed.ScratchValue{}, err
	}
	sv, err := b.scratch.plan(binding.AccountDataSlice{
		ValueTypeTag:         tag,
		AccountIndex:         uint8(dataIdx),
		Offset:               offset,
		ExpectedProgramOwner: uint8(ownerIdx),
	}, nil)
	if err != nil {
		return typed.ScratchValue{}, err
	}
	return b.push(sv), nil
}

func (b *LetBuilder) SplTokenAmount(account interface{}) (typed.ScratchValue, error) {
	i := b.AccountIndex(account)
	sv, err := b.scratch.PlanAtRemainingIndex(binding.SplTokenAccountAmount(0), uint8(i))
	if err != nil {
		return typed.ScratchValue{}, err
	}
	return b.push(sv), nil
}

func (b *LetBuilder) ClockSlot() (typed.ScratchValue, error) {
	sv, err := b.scratch.ClockSlot()
	if err != nil {
		return typed.ScratchValue{}, err
	}
	return b.push(sv), nil
}

func (b *LetBuilder) RentMinimumBalance(dataLen uint32) (typed.ScratchValue, error) {
	sv, err := b.scratch.RentMinimumBalance(dataLen)
	if err != nil {
		return typed.ScratchValue{}, err
	}
	return b.push(sv), nil
}

// Finish returns wire args and remaining accounts.
func (b *LetBuilder) Finish() FinishResult {
	return FinishResult{
		Args:      ToLetArgs(b.bindings),
		Bindings:  append([]typed.ScratchValue(nil), b.bindings...),
		Remaining: append([]typed.AccountMeta(nil), b.accounts...),
		Scratch:   b.scratch,
	}
}

// RemainingPubkeys converts remaining to solana public keys for ix.BuildLet.
func (r FinishResult) RemainingPubkeys() []solana.PublicKey {
	out := make([]solana.PublicKey, len(r.Remaining))
	for i, m := range r.Remaining {
		out[i] = solana.MustPublicKeyFromBase58(m.Pubkey)
	}
	return out
}

// BuildIx returns ifx_let for this batch.
func (b *LetBuilder) BuildIx() (solana.Instruction, error) {
	return b.scratch.IxLet(b)
}

File: rust-sdk/src/let_builder.rs
//! Multi-binding `ifx_let` batch with deduped remaining accounts.

use std::collections::HashMap;

use ifx_core::wire::{Expr, LetBinding, ValueType};
use solana_sdk::instruction::{AccountMeta, Instruction};
use solana_sdk::pubkey::Pubkey;

use crate::error::ScratchError;
use crate::ix::let_args_from_bindings;
use crate::scratch::FrameScratch;
use crate::typed::ScratchValue;

pub struct LetBuilderFinish<'a> {
    pub args: ifx_core::wire::LetArgs,
    pub bindings: Vec<ScratchValue>,
    pub remaining: Vec<AccountMeta>,
    pub scratch: &'a FrameScratch,
}

pub struct LetBuilder<'a> {
    scratch: &'a mut FrameScratch,
    accounts: Vec<AccountMeta>,
    index_by_key: HashMap<Pubkey, usize>,
    bindings: Vec<ScratchValue>,
}

impl<'a> LetBuilder<'a> {
    pub(crate) fn new(scratch: &'a mut FrameScratch) -> Self {
        Self {
            scratch,
            accounts: Vec::new(),
            index_by_key: HashMap::new(),
            bindings: Vec::new(),
        }
    }

    fn account_index(&mut self, account: Pubkey) -> u8 {
        if let Some(&idx) = self.index_by_key.get(&account) {
            return idx as u8;
        }
        let idx = self.accounts.len();
        self.index_by_key.insert(account, idx);
        self.accounts.push(AccountMeta {
            pubkey: account,
            is_signer: false,
            is_writable: false,
        });
        idx as u8
    }

    fn push(&mut self, sv: ScratchValue) -> ScratchValue {
        self.bindings.push(sv.clone());
        sv
    }

    fn plan_at(&mut self, binding: LetBinding, account: Pubkey) -> Result<ScratchValue, ScratchError> {
        let i = self.account_index(account);
        let sv = self
            .scratch
            .plan_at_remaining_index(binding, i)?;
        Ok(self.push(sv))
    }

    pub fn let_eval(&mut self, expr: Expr) -> Result<ScratchValue, ScratchError> {
        let sv = self.scratch.let_eval(expr)?;
        Ok(self.push(sv))
    }

    pub fn let_const_u64(&mut self, n: u64) -> Result<ScratchValue, ScratchError> {
        let sv = self.scratch.let_const_u64(n)?;
        Ok(self.push(sv))
    }

    pub fn let_const_bool(&mut self, v: bool) -> Result<ScratchValue, ScratchError> {
        let sv = self.scratch.let_const_bool(v)?;
        Ok(self.push(sv))
    }

    pub fn lamports(&mut self, account: Pubkey) -> Result<ScratchValue, ScratchError> {
        self.plan_at(LetBinding::AccountLamports { account_index: 0 }, account)
    }

    pub fn data_len(&mut self, account: Pubkey) -> Result<ScratchValue, ScratchError> {
        self.plan_at(LetBinding::AccountDataLen { account_index: 0 }, account)
    }

    pub fn let_account_key(&mut self, account: Pubkey) -> Result<ScratchValue, ScratchError> {
        self.plan_at(LetBinding::AccountKey { account_index: 0 }, account)
    }

    pub fn spl_token_amount(&mut self, account: Pubkey) -> Result<ScratchValue, ScratchError> {
        self.plan_at(LetBinding::SplTokenAccountAmount { account_index: 0 }, account)
    }

    pub fn spl_mint_decimals(&mut self, account: Pubkey) -> Result<ScratchValue, ScratchError> {
        self.plan_at(LetBinding::SplMintDecimals { account_index: 0 }, account)
    }

    pub fn clock_slot(&mut self) -> Result<ScratchValue, ScratchError> {
        let sv = self.scratch.clock_slot()?;
        Ok(self.push(sv))
    }

    pub fn rent_minimum_balance(&mut self, data_len: u32) -> Result<ScratchValue, ScratchError> {
        let sv = self.scratch.rent_minimum_balance(data_len)?;
        Ok(self.push(sv))
    }

    pub fn let_account_data_slice(
        &mut self,
        account: Pubkey,
        expected_owner: Pubkey,
        ty: ValueType,
        offset: u32,
    ) -> Result<ScratchValue, ScratchError> {
        let data_idx = self.account_index(account);
        let owner_idx = self.account_index(expected_owner);
        let sv = self.scratch.plan(
            LetBinding::AccountDataSlice {
                ty,
                account_index: data_idx,
                offset,
                expected_program_owner: owner_idx,
            },
            &[],
        )?;
        Ok(self.push(sv))
    }

    pub fn finish(&self) -> LetBuilderFinish<'_> {
        LetBuilderFinish {
            args: let_args_from_bindings(
                &self.bindings.iter().map(|v| v.binding.clone()).collect::<Vec<_>>(),
            ),
            bindings: self.bindings.clone(),
            remaining: self.accounts.clone(),
            scratch: self.scratch,
        }
    }

    pub fn build_ix(&mut self) -> Result<Instruction, ScratchError> {
        let fin = self.finish();
        self.scratch.ix_let_batch(&fin.bindings, &fin.remaining)
    }
}
